charger/delta.go

```diff
@@ -0,0 +1,277 @@
+package charger
+
+import (
+	"encoding/binary"
+	"fmt"
+	"time"
+
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/util"
+	"github.com/evcc-io/evcc/util/modbus"
+	"github.com/evcc-io/evcc/util/sponsor"
+)
+
+// Delta charger implementation
+type Delta struct {
+	conn *modbus.Connection
+	curr uint32
+}
+
+const (
+	//EV Charger Read Input Registers (0x04)
+	deltaRegState  = 100 // Charger State - UINT16 0: not ready, 1: operational, 10: faulted, 255: not responding
+	deltaRegCount  = 102 // EVSE Count - UINT16
+	deltaRegSerial = 110 // Charger Serial - STRING20
+	deltaRegModel  = 130 // Charger Model - STRING20
+
+	//EV Charger Write Multiple Registers (0x10)
+	deltaRegCommunicationTimeoutEnabled = 201 // Communication Timeout Enabled 0/1
+	deltaRegCommunicationTimeout        = 202 // Communication Timeout [s]
+	deltaRegFallbackPower               = 203 // Fallback Power [W]
+
+	//EVSE Read Input Registers (0x04)
+	deltaRegEvseState                 = 1000 // EVSE State - UINT16 0: Unavailable, 1: Available, 2: Occupied, 3: Preparing, 4: Charging, 5: Finishing, 6: Suspended EV, 7: Suspended EVSE, 8: Not ready, 9: Faulted
+	deltaRegEvseChargerState          = 1001 // EVSE Charger State - 0: Charging process not started (no, vehicle connected), 1: Connected, waiting for release (by, RFID or local), 2: Charging process starts, 3: Charging, 4: Suspended (loading paused), 5: Charging process successfully com-, pleted (vehicle still plugged in), 6: Charging process completed by, user (vehicle still plugged in), 7: Charging ended with error (vehicle, still connected)
+	deltaRegEvseActualOutputVoltage   = 1003 // EVSE Actual Output Voltage [V]
+	deltaRegEvseActualChargingPower   = 1005 // EVSE Actual Charging Power [W]
+	deltaRegEvseActualChargingCurrent = 1005 // EVSE Actual Charging Current [A]
+
+	deltaRegEvseChargingTime  = 1017 // EVSE Charging Time [s]
+	deltaRegEvseChargedEnergy = 1019 // EVSE Charged Energy [Wh]
+
+	deltaRegEvseCurrentPowerConsumptionL1 = 1049 // EVSE Current Power Consumption L1 (grid) [W]
+	deltaRegEvseCurrentPowerConsumptionL2 = 1051 // EVSE Current Power Consumption L2 (grid) [W]
+	deltaRegEvseCurrentPowerConsumptionL3 = 1053 // EVSE Current Power Consumption L3 (grid) [W]
+
+	//EVSE Write Multiple Registers (0x10)
+	deltaRegEvseChargingCurrentLimit = 1600 // EVSE Charging Power Limit - UINT32 [W]
```

Comment:
@mirgonet soweit ich sehe lässt sich der Ladestrom steuern (oder ist das trotz des Namens tatsächlich ein Power Limit)? Falls sich der Strom steuern ließe könnte die ganze Power-Magie verschwinden. 

Darauf würde ich hier in jedem Fall verzichten da das eigentlich schon der Ladepunkt macht. Dafür müsstest Du nur `loadpoint.Controller` implementieren und bekämst damit Zugriff auf `lp.GetPhases()`.

Reply:
Bin anfänglich auch davon ausgegegangen, dass der Ladestrom irgendwie limitiert werden kann. Leider gibt die mir vorliegende Doku nur ein einziges Register zur Leistungssteuerung her und das läuft in Watt. Wäre mir auch lieber den Workaround zu vermeiden.

Reply:
So, habe den Workaround durch `lp.GetPhases()` ersetzt. Danke für den Hinweis. Werde es morgen mal testen, aktuell ist der Akku leider voll und ich kann nicht laden.

Bin auch noch am überlegen ob ich den Status vom `State` oder `ChargerState` Register ableite und ob ich das Enable wie gehabt über setCurrent = 0 oder das `SuspendCharging` Register implementiere.

Reply:
Hab jetzt erstmal die Variante "Leistung 0" gewählt.
Mal sehen wie sich das in der Realität verhält...